client/app/components/dashboards/DashboardGrid.jsx

```diff
@@ -33,6 +33,45 @@ const WidgetType = PropTypes.shape({
 const SINGLE = "single-column";
 const MULTI = "multi-column";
 
+const DashboardWidget = React.memo(function DashboardWidget({
```

Comment:
I assume the performance issue isn't with the `RestrictedWidget` or `TextboxWidget` (correct me if I wrong), but with visualization rendering. Maybe we should actually memoize one of the more basic components down the stream? This will later be useful in other places we render visualizations (query page).

Reply:
Why not both? :grin:

I actually wanted to memoize as high as possible in the tree, thus shaving off more time.

Reply:
Because memoization can have unwanted side effects that we rather avoid. The assumptions I'm working with are: 1) for the most part re-rendering is a cheap operation in React; 2) in many cases even when you re-render React won't in fact do anything because the DOM stays the same.

For this reason I would rather keep memoization for places it's really needed.

If you prefer, we can merge this as is and explore memoization of a different component as a follow up.

Reply:
That's kinda of why I wanted to have more testing time to it, but if it ended up being safe to memoize in an Upper level, there should be no problem to memoize in a lower one as well.

From my understanding it's a matter of being able to determine a component from its props/state, so if the tree was somehow composed by immutable objects (or basic variable props), it should be fine to use `React.memo`.

But I do understand your concern and also share it, that's the reason I was targeting #4724 and #4726 to v9 and left away this one. Since this PR is targeted to #4724 and not master, and I am using it for results comparison (mainly for alternatives we are discussing in #4724), let's leave this PR here and I'll compare how it differs from memoizing the visualization.

Reply:
As I imagined, memoizing the Visualization instead of the whole widget helps, but not quite as much:

| Dashboard | master | Memoized Widgets | Memoized Visualizations |
| --------------- | --------- | ---------- | ------------------------- |
| Dashboard for testing React migration |  38s (~) | 20s (-47%) | 30s (-21%) |

---

What happens is that it doesn't really seem that are the visualizations that are slowing down the dashboards, but it rather seem a combination of: 
1. re-render numbers; and
2. number of components down the stream. 

The re-render process is mostly a cheap operation, as you said (<1ms), but it currently scales.

In brief, #4724 is an attempt to solve 1 and this PR is an attempt to solve 2.

---

A few screenshots to illustrate how memoizing visualizations differs from memoizing widgets:
![Screenshot from 2020-03-17 12-18-49](https://user-images.githubusercontent.com/3356951/76871223-929ebf80-6849-11ea-907b-c00c1e9d49be.png)
![Screenshot from 2020-03-17 12-35-32](https://user-images.githubusercontent.com/3356951/76872889-ce3a8900-684b-11ea-8a44-eaf1f4980b24.png)


Basically re-rendering a Visualization seems not that expensive as well, it mostly helped here because we reduced part of 2.